### ISSUE TITLE
Is there the set_proxy API for the Node SDK, similar to the Azure_SDK for Python - set_proxy()

### ISSUE THREAD

**Issue by NNM11** [Azure/azure-sdk-for-node#1622]
The main problem is how-to to configure proxy with the SDK to access Azure cloud.  

The Azure_SDK for Python had the set_proxy() API for the ServiceBus to support direct HTTP tunneling.  But the Azure_SDK for Node did not have this API.  

Is the proxy support not available for the Azure_SDK for Node?  Shouldn't all Azure_SDKs (C, Ruby, Node, Python, etc). have the same supports/apis.

Also, most business client devices will be behind some firewall when communicates to an Azure Cloud service.  The proxy support for direct tunneling is essential for the Azure_SDK for Node to be used in these scenarios.

**Comment by amarzavery** [comment]
https://github.com/Azure/azure-sdk-for-node/blob/master/lib/common/lib/services/serviceclient.js#L838 
Line 838 to 860 is what you would want to look at. 
Every client is a child of the ServiceClient.

**Comment by NNM11** [comment]
Thanks for the pointer.  It looked like the ServiceClient code automatically set up proxy for all clients using the environment HTTP_PROXY by default.

**Comment by amarzavery** [comment]
Closing this issue. Feel free to reopen if required.

**Comment by NNM11** [comment]
It was too good to be true!  
I set the environment variable HTTPS_PROXY=http://<myproxy>:411 and HTTP_PROXY=http://<myproxy>:412.
then run the azure ServiceBus Client code.   It failed with error 'tunneling socket could not be established, cause=socket hang up'
Question:
- Is there any thing else required for the proxy configuration?  header, auth, path, TTL, etc.

**Comment by amarzavery** [comment]
Can you provide a sample of your script of what you are trying to achieve? 
Which package are you using: "azure-sb" (data plane) or "azure-asm-sb" (mgmt. plane)
- If you are using the data plane package then the final parent class is [ServiceClient.js](https://github.com/Azure/azure-sdk-for-node/blob/master/lib/common/lib/services/serviceclient.js).
- If you are using the mgmt. plane then the final parent class is [Service.js](https://github.com/Azure/azure-sdk-for-node/blob/master/lib/common/lib/services/service.js)

Both the parent classes have their version of _setDefaultProxy() method which is the last step in the constructor. That should give you some idea.

If you can provide you sample script, may be I can try to help you further.

**Comment by NNM11** [comment]
Hi,

I used the azure-sb package.  I had the 'https_proxy=http://proxy-server.xyz.com:412' environment set up.

my script is:
"
var sbConnectionStr='<the full SB SAS from Azure>';
var serviceBusObj = azure.createServiceBusService(sbConnectionStr);

if (process.env.HTTPS_PROXY) {
	// create basic credential for proxy
	var username = 'username';
	var password = 'password';
	var auth = 'Basic ' + new Buffer(username + ':' + password).toString('base64');
	
	// get the default proxy created from the ServiceClient constructor & modify it.
	var defaultProxy = serviceBusObj.proxy;
	// update the proxy object
	defaultProxy.auth = auth;
	// override the default proxy - #838  / lib / common / lib / services / serviceclient.js 
	serviceBusObj.setProxy(defaultProxy);
}

// Create the SB Topic
var topicOptions = {
    "DefaultMessageTimeToLive" : "PT24H",				// message TTL in Topic is 24H
    "EnableDeadLetteringOnMessageExpiration" :  true 	//
};
    // Create a new topic if not existed.
    serviceBusObj.createTopicIfNotExists('myDevice', topicOptions , function (error) {
		// ignore error 409 - topic existed
		if (error && (error.statusCode != 409)) {
			console.log('ERR - Failed to create topic for device' + devTopic + ' - ' + error);
			if (_debug > 1) { console.log('*** DBG: Proxy: ' + JSON.stringify(serviceBusObj.proxy) + '\n');}
		    return;
        }
    });
"
** The createTopicsIfNotExists() returned error
ERR - Failed to create topic for device myDevice - Error: tunneling socket could not be established, cause=socket hang up
*** DBG: Proxy: {"protocol":"http:","slashes":true,"auth":null,"host":"proxy-server.xyz.com:412","port":"412","hostname":"proxy-server.xyz.com","hash":null,"search":null,"query":null,"pathname":"/","path":"/","href":"http://proxy-server.xyz.com:412/"}

** From my previous exp working with proxy, the proxy options such as: auth, headers {Proxy-Authorization, host, path}, path, etc. must be filled in to establish the outbound connection.  I assume they are filled in by the azure SDK request module implementer - _performRequest #190.   I overrode the 'auth' just to provide the basic auth.  The debug proxy used by the request was not completely configured!?    

** BTW - Where does the socket tunneling handle?  and is there a proxy test to verify that the proxy configurations are valid?

**Comment by NNM11** [comment]
Notes:
- I took the TEST app from 'azure-sdk-for-node\test\services\serviceBus\sbtopic.js' and used the default HTTPS_PROXY with a test ServiceBus SAS.  It did not work - failed on the same error.  
- The same code ran fine without proxy.  
- I had other node app called to a cloud server's REST API, using HTTPS_PROXY.  I was able to establish the tunnel with the same proxy server with some basic proxy options.  

This seems to indicate that the Azure SDK cannot support proxy with just the default proxy URL - 'http://proxy-server:port'.
